src/Csv/CsvProvider.fs

```diff
@@ -75,15 +75,26 @@ type public CsvProvider(cfg:TypeProviderConfig) as this =
 
       using (IO.logTime "TypeGeneration" sample) <| fun _ ->
 
-      let csvType, csvErasedType, stringArrayToRow, rowToStringArray = 
+      let csvType, csvErasedType, rowType, stringArrayToRow, rowToStringArray = 
         inferredFields 
         |> CsvTypeBuilder.generateTypes asm ns typeName (missingValuesStr, cultureStr) replacer 
-  
+
+
+      let stringArrayToRowVar = Var("stringArrayToRow", stringArrayToRow.Type)
+      let rowToStringArrayVar = Var("rowToStringArray", rowToStringArray.Type)
+      
+      let paramType = (replacer.ToRuntime typedefof<seq<_>>).MakeGenericType(rowType)
+      let headers = match sampleCsv.Headers with 
+            | None -> <@@ None: string[] option @@> 
+            | Some headers -> Expr.NewArray(typeof<string>, headers |> Array.map (fun h -> Expr.Value(h)) |> List.ofArray) |> (fun x-> <@@ Some (%%x : string[]) @@>)
+      let newfn = ProvidedMethod("New", [ ProvidedParameter("rows", paramType) ], csvType, IsStaticMethod = true, InvokeCode = (fun (Singleton paramValue) ->
```

Comment:
Thinking a bit more about it, I think it's preferable to use a constructor instead of a static method New, for better parity with the other providers